Improve cops by using TopLevelGroup

pirj opened the pull request:
The previously used `TopLevelDescribe` suffered a few flaws:
 - it was slow: https://github.com/rubocop-hq/rubocop-rspec/pull/925#discussion_r436988718
 - it ignored non-`describe` top-level example groups https://github.com/rubocop-hq/rubocop-rspec/pull/925#issuecomment-641140259

This pull request fixes those problems by using a newly introduced `TopLevelGroup`.
Also fixes a flaw in `TopLevelGroup` when example groups wrapped inside `module` or `class` were missed.

NOT DONE: I'm on the fence if we should deprecate `TopLevelDescribe` before removing it. We most probably don't have much releases left before 2.0, and the majority of people will jump straight to 2.0 skipping that pre-2.0 release. We may add this pull request (and others that replace usages of `TopLevelDescribe` for `TopLevelGroup`) as reference examples for migration.

Fixes #948

---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Updated documentation.
* [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

pirj commented on lib/rubocop/cop/rspec/file_path.rb:
```diff
@@ -57,34 +57,44 @@ module RSpec
-            node,
-            message: format(MSG, suffix: glob)
-          )
+          ensure_correct_file_path(send_node, described_class, arguments)
```
Code climate insisted to break this apart.

pirj commented on spec/rubocop/cop/rspec/multiple_describes_spec.rb:
```diff
@@ -3,28 +3,36 @@
+  it 'flags shared example groups' do
+    expect_offense(<<-RUBY)
+      shared_examples_for 'behaves' do
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Do not use multiple top-level example groups - try to nest them.
```
This is a bigger change, but I believe a good one.

Darhazer replied:
I think it's quite common to define multiple shared examples in a single, helper file?

pirj replied:
Agree, will narrow it back down to example groups only 👍

We may think of this option later on.

pirj commented on spec/rubocop/cop/rspec/nested_groups_spec.rb:
```diff
@@ -35,11 +54,32 @@
     expect(cop.config_to_allow_offenses[:exclude_limit]).to eq('Max' => 4)
   end
 
-  it 'ignores non-spec context methods' do
```
It wasn't ignoring. Just the example didn't have enough nesting to trigger the offence.

Darhazer commented on lib/rubocop/cop/rspec/describe_class.rb:
```diff
@@ -51,13 +51,21 @@ class DescribeClass < Base
+        def on_top_level_group(top_level_node)
+          node, described = described(top_level_node)
+
+          return unless described
```
Can we have `!nil?` in the matcher and skip this check

pirj replied:
I keep forgetting about the yielding feature of a node pattern.

This comment of yours, however, led to a major refactoring 😆

Darhazer commented on lib/rubocop/cop/rspec/describe_method.rb:
```diff
@@ -17,16 +17,24 @@ module RSpec
 
-          add_offense(second_arg)
+        def on_top_level_group(node)
+          second_argument = second_argument(node)
```
we implicitly check that this is a `describe` this way. I'm not strictly against it, but since some cops need to operate only on describe, I wonder if we can't customize the on_top_level_group to actually have separate callbacks for on_top_level_describe. We still use the same code, just call different method if it is available and the top level group is a describe. And we don't make any breaking change in the API

pirj replied:
Implemented separate hooks, `on_top_level_example_group` and `on_top_level_group` 👍

Darhazer replied:
We still check implicitly if this is a describe, by using the second_argument matcher

pirj commented on lib/rubocop/cop/rspec/describe_class.rb:
```diff
@@ -22,49 +22,38 @@ module RSpec
+              !(send #{RSPEC} :describe ... (hash <#rails_metadata? ...>)) ]
+            ...
+          )
+        PATTERN
```
Collapsed several significantly overlapping node patterns into one.
Moved guard statements into node pattern as well.

There still seems to be some redundancy in this node pattern, similarly to `SubjectStub`:
```
          {
            (block (send nil? :subject (sym $_)) args ...)
            (block (send nil? $:subject) args ...)
          }
```

I'll open a ticket for `rubocop-ast`.

Darhazer replied:
Is `[ ...]` a new syntax for `{ ... }`? Does it work on the minimum supported rubocop version or should we bump the requirements?

pirj replied:
`[ ]` is logical "and", it's around for a while, just not too frequently used.

I find it too hard to reason about boolean login, this node pattern sounds like:
a block
  with a send of `describe` to `RSpec`/nil?
    where the first argument is
      neither a const not a string resembling a const
      *and* the rest of the arguments don't contain a hash with Rails-like metadata

Darhazer replied:
I'm confused by the `(send #rspec? :describe ` part in the second argument of `[ ]`

pirj replied:
@Darhazer Why?

It's basically the same as previously (which I forgot to remove during merge resolution):
```ruby
        def_node_matcher :describe_with_rails_metadata?, <<-PATTERN
          (send #rspec? :describe !const ...
            (hash <#rails_metadata? ...>)
          )
        PATTERN
```

`!const` check is not necessary anymore, it's checked in the first part of `[ ]`.

bquorning replied:
I too find this node pattern a bit cryptic. Perhaps a bit more duplication can be justified if it makes this code easier to read?

pirj replied:
Good point guys. Split this node pattern into two parts. Looks more readable now.

marcandre replied:
FYI the shiny newly released `rubocop-ast` allows comments in patterns

pirj commented on lib/rubocop/cop/rspec/describe_method.rb:
```diff
@@ -17,16 +17,24 @@ module RSpec
+          second_argument = second_argument(node)
+
+          return unless second_argument
+          return if second_argument.str_content.start_with?('#', '.')
```
I'm ok to turn this into a method and pushing this into the node pattern, but decided not to fix everything in one go.

pirj commented on lib/rubocop/cop/rspec/file_path.rb:
```diff
@@ -57,34 +57,44 @@ module RSpec
         def_node_search :routing_metadata?, '(pair (sym :type) (sym :routing))'
 
-        def on_top_level_describe(node, args)
-          return unless const_described?(node) && single_top_level_describe?
```
That was the only remaining usage of `single_top_level_describe?`, and I decided to inline it.

pirj commented on lib/rubocop/cop/rspec/file_path.rb:
```diff
@@ -57,34 +57,43 @@ module RSpec
-            node,
-            message: format(MSG, suffix: glob)
-          )
+            ensure_correct_file_path(send_node, described_class, arguments)
```
Code climate insisted on this extraction.

pirj commented on lib/rubocop/rspec/top_level_group.rb:
```diff
@@ -10,11 +10,21 @@ module TopLevelGroup
-      def on_block(node)
-        return unless respond_to?(:on_top_level_group)
-        return unless top_level_group?(node)
+      def on_new_investigation
```
This is a hair more efficient than `on_block`.

pirj commented on lib/rubocop/rspec/top_level_group.rb:
```diff
@@ -10,11 +10,21 @@ module TopLevelGroup
+      end
+
+      def top_level_groups
+        return [] unless root_node
```
To satisfy our "empty spec" smoke test. Is there a better way around this?

Darhazer replied:
return from the `on_new_investigation` if there is no ast?

pirj replied:
Good idea, thanks!

pirj replied:
Fixed.

pirj commented on lib/rubocop/rspec/top_level_group.rb:
```diff
@@ -23,16 +33,13 @@ def top_level_group?(node)
+      def top_level_nodes(node)
+        if node.begin_type?
+          node.children
+        elsif node.module_type? || node.class_type?
```
Previously,
```ruby
module A
  class B
    RSpec.describe 'C' do
```
was missed.

pirj commented on spec/rubocop/cop/rspec/nested_groups_spec.rb:
```diff
@@ -20,6 +20,36 @@
+          context 'when bar' do
+            context 'when baz' do
+              context 'when qux' do
+              ^^^^^^^^^^^^^^^^^^ Maximum example group nesting exceeded [4/3].
```
`shared_examples_for` is not counted.
If we count it in, even one of our own specs triggers an offence:
```ruby
RSpec.describe RuboCop::Cop::RSpec::PredicateMatcher, :config do
  context 'when enforced style is `explicit`' do
    shared_examples 'explicit common' do
      context 'when custom matchers are allowed' do
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Maximum example group nesting exceeded [4/3].
```

pirj commented:
Ping @bquorning @Darhazer

Darhazer commented on lib/rubocop/rspec/top_level_group.rb:
```diff
@@ -10,11 +10,21 @@ module TopLevelGroup
+      def on_new_investigation
+        top_level_groups.each do |node|
+          if respond_to?(:on_top_level_example_group)
+            example_group?(node, &method(:on_top_level_example_group))
```
and we can have the same for other methods like describe

pirj replied:
Yep, if we'll eventually need them.

Initially, I was going to add code that would call `on_top_level_shared_group`, but that would only count one usage, so decided not to add it just now.

pirj commented:
Simplified TopLevelGroup by introducing dummy methods.
I've removed `TopLevelGroup`'s checks if those methods exist in the including class, since of `TopLevelGroup` is included, that means that one of those methods is overridden. Hence, we can use dummy methods and call them instead of checking if they exist in the consumer.

pirj commented:
@bquorning @Darhazer ping

pirj commented on lib/rubocop/rspec/top_level_group.rb:
```diff
@@ -10,28 +10,31 @@ module TopLevelGroup
-      def single_top_level_group?
-        top_level_groups.one?
+      def top_level_groups
+        @top_level_groups ||=
```
I was following [this note](https://github.com/rubocop-hq/rubocop-rspec/pull/932#discussion_r439636213):

> Right now a new cop instance is created for every file

There might have been some changes after the doc has been introduced.

As a quick experiment, I've set a breakpoint in random cop's `initialize` and ran `rubocop`:
```
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
From: /Users/pirj/source/rubocop-rspec/lib/rubocop/cop/rspec/described_class.rb:63 RuboCop::Cop::RSpec::DescribedClass#initialize:

    61: def initialize(*args)
    62:   require 'pry'; binding.pry
 => 63:   super
    64: end

[1] pry(#<RuboCop::Cop::RSpec::DescribedClass>)>
.
```

@marcandre Can you please shed light on this? Isn't this doc misleading?

I used to think it would be quite hard to reuse cop instances since depending on the directory they may receive different config (due to directory-local `.rubocop.yml`).

pirj replied:
The [cop mentioned in the referenced discussion, `SurroundingSpace`](https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/mixin/surrounding_space.rb#L59) does reset its state:
```ruby
      def on_new_investigation
        @token_table = nil
```

pirj replied:
What's even more strange:
```ruby
        def on_new_investigation
          print "N"
          super
        end

        def initialize(*args)
          print "#"
          super
        end
```
outputs:
```
##.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#C#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#N.#.
```

🤔

marcandre replied:
@pirj Currently, cops being persisted [is opt-in](https://docs.rubocop.org/rubocop/0.88/v1_upgrade_notes.html#cop-persistence). That being said, I'm wondering if a cop that implements `on_new_investigation` shouldn't be persisted by default.
BTW, different config => new cop, but that's a very rare setup. 
It looks like the cop is initialized a bunch of times before doing the actual run, that seems like a bug somewhere, I'll check it out.

pirj replied:
I'm looking at default cop's initializer and struggling to understand if there are any performance/memory usage improvements of reusing cop instances:
```ruby
      def initialize(config = nil, options = nil)
        @config = config || Config.new
        @options = options || { debug: false }
        reset_investigation
      end

      def reset_investigation
        @currently_disabled_lines = @current_offenses = @processed_source = @current_corrector = nil
      end
```

It would make sense if the expensive `Config.new` would be called on each extra cop initialization, but it isn't:
```ruby
      # Called between investigations
      def ready
        return self if self.class.support_multiple_source?

        self.class.new(@config, @options) # <== `@config` is reused
      end
```

pirj replied:
Please don't get me wrong, I don't really mind (as a performance-last person) to remove memoization here. We're only calling this `top_level_groups` method once per top-level example group. And those who have more than one, probably deserve to slightly suffer anyway.

I'm more concerned about the confusion that this reuse might deal if once this option will become opt-out rather than opt-in. Cop specs won't be bringing memoization to the light, will they? With all numerous well-unknown RuboCop extensions this can have quite an impact.

marcandre replied:
Nothing major indeed. I wanted to speedup the callback cache building in `Commissionner`, but it could be done another way (per class instead of per instance). Some cops *want to* persist, though (there was an issue on that), which is why it's possible to do so.

pirj replied:
Also, speaking of per-directory configs. In theory, it's allowed, and I don't see a reason to disrespect this option.

In the practice of working with majestic monoliths composed of Rails engines, in-repo gems and libraries in git submodules it's totally not uncommon to have different styles and different configuration, at least in some directories.

pirj replied:
> Some cops want to persist

Got it 👍 That's reasonable, I can imagine some use cases for that.

However, does it make sense to keep this notice for the rest of the cops that don't want and won't persist by default? https://github.com/rubocop-hq/rubocop/blob/6b36c4490c2d7c41349e4f71721bf0cf91920578/docs/modules/ROOT/pages/v1_upgrade_notes.adoc#if-your-class-uses-instance-variables

marcandre replied:
> Cop specs won't be bringing memoization to the light, will they? With all numerous well-unknown RuboCop extensions this can have quite an impact.

Yes, that was my thinking too.

> I don't see a reason to disrespect this option.

Absolutely :-) 

> However, does it make sense to keep this notice for the rest of the cops

You mean the recommendation to clear the data with `on_new_investigation`? I guess it's not necessary. I should also be clearer reword the "will be called on different if" to "will not be called unless"...

pirj replied:
I would remove the whole "If your class uses instance variables" section, as it's more confusing than helpful considering the current "one source" - "one cop instance" behaviour.

Even for those cops that need to be persisted, I can think of ~global variables~ class variables to persist the data between the runs.

By the way, there's one more thing that can impact those persisted cops - `--parallel`. Do we parallelize by cops or by source?

marcandre replied:
=> https://github.com/rubocop-hq/rubocop/pull/8440

marcandre replied:
> Even for those cops that need to be persisted, I can think of class variables to persist the data between the runs.

Not quite... See https://github.com/rubocop-hq/rubocop/issues/7968 if interested

> By the way, there's one more thing that can impact those persisted cops - --parallel. Do we parallelize by cops or by source?

I imagine it's by source, but we could segregate those cops that need it; I suspect they simply run them without `--parallel` though.

pirj replied:
@bquorning As a bottom-line: I wouldn't worry about `@top_level_groups ||=` used here unless any of the cops would decide to opt-in for reusage *and* use `TopLevelGroup` which I highly doubt.

pirj commented on lib/rubocop/cop/rspec/file_path.rb:
```diff
@@ -57,34 +57,42 @@ module RSpec
 
-          glob = glob_for(args)
+          const_described(node) do |send_node, described_class, arguments|
+            next if routing_spec?(arguments)
```
At the point when we spawn `rubocop-rspec-rails`? :D

pirj commented on lib/rubocop/rspec/top_level_group.rb:
```diff
@@ -10,28 +10,31 @@ module TopLevelGroup
+        top_level_groups.each do |node|
+          example_group?(node, &method(:on_top_level_example_group))
+          on_top_level_group(node)
+        end
```
Why?
https://github.com/rubocop-hq/rubocop/blob/6b36c4490c2d7c41349e4f71721bf0cf91920578/lib/rubocop/cop/base.rb#L67:
```ruby
      # Called before all on_... have been called
      # When refining this method, always call `super`
      def on_new_investigation
        # Typically do nothing here
      end
```

pirj replied:
I've answered my own question 🤦 

> When refining this method, always call `super`

pirj commented on lib/rubocop/rspec/top_level_group.rb:
```diff
@@ -25,14 +25,16 @@ def top_level_group?(node)
-        if root_node.begin_type?
-          root_node.children
+      def top_level_nodes(node)
+        if node.begin_type?
```
I don't really have a strong preference here, but
```ruby
        case node.type
        when :begin
          node.children
        when :module, :class
          top_level_nodes(node.body)
        else
          [node]
        end
```
feels more like poking around in internals.

pirj commented:
Loud thinking

```ruby
describe Weather do
  let(:temp) { 20.celsius }
  context "when it's warm" do
    let(:temp) { 30.celsius } # <== should it count as an additional `let`, or should it collapse with an existing one?
```

marcandre commented:
> Loud thinking
> 
> ```ruby
> describe Weather do
>   let(:temp) { 20.celsius }
>   context "when it's warm" do
>     let(:temp) { 30.celsius } # <== should it count as an additional `let`, or should it collapse with an existing one?
> ```

I assumed it was one. It makes even less sense to me to count the repeated `let`.
